[Law Wei Ming] iP

lawwm opened the pull request:
### DukePro
> “Action is the foundational key to all success.” – Pablo Picasso ([source](https://dansilvestre.com/productivity-quotes))

DukePro is a task management program to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/lawwm/ip/releases/tag/A-Jar)
2. run the jar file.
3. add your tasks.
4. let it manage your tasks for you 👍 💯 🧠  

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon!)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level-1: Greet, Echo, Exit

Commit:
Level-2

Commit:
Level-3: Mark as Done

Commit:
Level-4: ToDos, Events, Deadlines

Commit:
Level-6

Commit:
Level-5

Commit:
branch-level-7

Commit:
Level-8: Added time and date

Commit:
Merged branch-level-7 into master

Commit:
Merged level-8-branch with master

Commit:
A-MoreOOP

Commit:
A-Packages

Commit:
A-JUnit

Commit:
Branch-level-9 Find

Commit:
Added JavaDoc to program

Commit:
Applied coding standard changes

Commit:
Fixed merge conflicts between javadoc branch and find branch

Commit:
Merged with branch-A-CodingStandard

Commit:
Merge branch 'branch-A-Gradle' into branch-Gradle

Commit:
Set up Gradle

Commit:
Gradle set up

Commit:
Altered runtest.bat to fit new changes

Commit:
Changed enum spacing

Commit:
Add complete text-ui-text to git

Commit:
Add GUI to Duke

Commit:
Changed GUI profile picture

Commit:
Merge gradle branch with master

tandeshao commented on src/main/java/DeadLine.java:
```diff
@@ -3,7 +3,6 @@
 public class DeadLine extends Task {
-    protected String by;
     protected LocalDate bydate;
     protected LocalTime bytime;
```
The variable names should be changed to something more understandable and descriptive.

lawwm replied:
Thanks, I missed out on that.

tandeshao commented on src/main/java/Duke.java:
```diff
@@ -1,234 +1,53 @@
-import java.awt.*;
```
Remove the wildcard import and import the packages one by one.

tandeshao commented on src/main/java/Duke.java:
```diff
@@ -1,234 +1,53 @@
-            case FAREWELL: reply = "Bye. Hope to see you again soon!\n";
-                break;
-            default: reply = "Are you finally done?";
-                break;
```
What a clean and succinct way to check for the different conditions! I am impressed 😄 !

tandeshao commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,57 @@
+            default: style = "";
+                break;
+        }
+        return style;
```
Looks good!

tandeshao reviewed (commented):
Overall your code is readable and looks good! It follows the coding standard very closely 😄!

Jacky142857 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,78 @@
+    protected String speak(Dialogue option, Integer num) {
+        String reply;
+        switch (option) {
+            case NUMLEFT: reply = "Now you have " + num.toString() + " task in the list\n";
```
According to coding standard, there should be no indentation for case clauses.

Jacky142857 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,78 @@
+    protected String speak(Dialogue option) {
+        String reply;
+        switch (option) {
+            case GREETING:
```
wrong indentation for case clauses

lawwm replied:
Good catch! I'll be changing that.

Jacky142857 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,90 @@
+    public abstract String serialize();
+
+
+    public boolean contains(String keyword) {
```
Maybe can name this method as 'hasKeyword'. Though I think 'contains' is intuitive enough.

Jacky142857 reviewed (commented):
The code is clean and well written. The variables are named in a specific and easy to understand way. I've also learnt some good implementations from you such as overwriting the equals method. You did a good job, well done!

Commit:
Add sorting feature to Duke

There is no sorting feature currently, and
it needs to be implement for user friendliness.

There are comparators stored within enum, and
user's choice of keywords decides the comparator
to be used.

Commit:
Add background and widen scrollpane for GUI

Commit:
Improve code quality

Apply Single Level of Abstraction Principle
for Command class and subclasses for formatting reply.

Add guard clauses for the load method of Storage class,
to make happy path more prominent.

Commit:
Add assertions to Duke

No checks for state currently.

Add assertion to check for state of
program and ensure correctness.

It is done to test my assumptions, and
unlike test units, it is included in
production code.

lawwm committed:
Merge pull request #2 from lawwm/branch-A-CodeQuality

Improve code quality

Commit:
Fix assertion merge conflicts

lawwm committed:
Merge pull request #1 from lawwm/branch-A-Assertions

Add assertions to Duke

Commit:
Use stream within TaskList

Commit:
Use stream to serialize list

Commit:
Fix issues from code review

There are coding standard issues with previous commits.

Therefore, in order to fit with the standardised coding standard,
change the code to fit coding standard. Alter the switch cases within
Command.java as well as inappropriate method and variable names.

This is so all code will look as if they were written by
a single developer.

Commit:
Refactor code slightly

Commit:
Finalize features

Commit:
Add credit statement

Commit:
Add user guide

Commit:
Update user guide

Commit:
Fix formatting error

Commit:
Remove line to fix github pages formatting

Commit:
Remove redundant line in user guide

Commit:
Fix issues found in nus-se-bot analysis
